src/macaron/parsers/github_workflow_model.py

```diff
@@ -0,0 +1,426 @@
+    schedule: NotRequired[list[ScheduleItem]]
+
+
+class Container(TypedDict):
+    image: str
+    credentials: NotRequired[Credentials]
+    env: NotRequired[Env]
+    ports: NotRequired[list[float | str]]
+    volumes: NotRequired[list[Volume]]
+    options: NotRequired[str]
+
+
+Permissions = Union[Literal["read-all", "write-all"], PermissionsEvent]
+
+
+ReusableWorkflowCallJob = TypedDict(
+    "ReusableWorkflowCallJob",
+    {
+        "name": NotRequired[str],
+        "needs": NotRequired[JobNeeds],
+        "permissions": NotRequired[Permissions],
+        "if": NotRequired[Union[bool, float, str]],
+        "uses": str,
+        "with": NotRequired[Env],
+        "secrets": NotRequired[Union[Env, Literal["inherit"]]],
+        "strategy": NotRequired[Strategy],
+        "concurrency": NotRequired[Union[str, Concurrency]],
+    },
+)
+
+
+NormalJob = TypedDict(
+    "NormalJob",
+    {
+        "name": NotRequired[str],
+        "needs": NotRequired[JobNeeds],
+        "permissions": NotRequired[Permissions],
+        "runs-on": Union[str, list, RunsOn, StringContainingExpressionSyntax, ExpressionSyntax],
+        "environment": NotRequired[Union[str, Environment]],
+        "outputs": NotRequired[dict[str, str]],
+        "env": NotRequired[Env],
+        "defaults": NotRequired[Defaults],
+        "if": NotRequired[Union[bool, float, str]],
+        "steps": NotRequired[list[Step]],
+        "timeout-minutes": NotRequired[Union[float, ExpressionSyntax]],
+        "strategy": NotRequired[Strategy],
+        "continue-on-error": NotRequired[Union[bool, ExpressionSyntax]],
+        "container": NotRequired[Union[str, Container]],
+        "services": NotRequired[dict[str, Container]],
+        "concurrency": NotRequired[Union[str, Concurrency]],
+    },
+)
+
+
+Workflow = TypedDict(
+    "Workflow",
+    {
+        "name": NotRequired[str],
+        "on": Union[Event, list[Event], On],
+        "env": NotRequired[Env],
+        "defaults": NotRequired[Defaults],
+        "concurrency": NotRequired[Union[str, Concurrency]],
+        "jobs": dict[str, Union[NormalJob, ReusableWorkflowCallJob]],
+        "run-name": NotRequired[str],
+        "permissions": NotRequired[Permissions],
+    },
+)
+
+
+from dataclasses import dataclass
+
+# MODIFIED: Extensions to add convenience types and cast operators
+from typing import Generic, TypeVar, cast
+
+Job = NormalJob | ReusableWorkflowCallJob
+RunStep = Step5
+ActionStep = Step4
+
+T = TypeVar("T")
+
+
+@dataclass
+class Identified(Generic[T]):
+    id: str
+    obj: T
+
+
+def as_normal_job(job: Job) -> NormalJob | None:
+    if "uses" in job:
+        return None
+    return cast(NormalJob, job)
+
+
+def as_reusable_workflow_call_job(job: Job) -> ReusableWorkflowCallJob | None:
+    if "uses" in job:
+        return cast(ReusableWorkflowCallJob, job)
+    return None
+
+
+def as_run_step(step: Step) -> RunStep | None:
+    if "run" in step:
+        return cast(RunStep, step)
+    return None
```

Comment:
This is one way to do this kind of type-casting and I think it's fine.

I'm more used to using [type guards](https://peps.python.org/pep-0647/) in situations like these. Here for example, the type guard function can be:

```python3
def is_run_step(step: Step) -> TypeGuard[RunStep]:
    return "run" in step
```

in which case the following function

```python3
def get_run_step(step: Step) -> str | None:
    run_step = as_run_step(step)
    if run_step is not None:
        return run_step["run"]
    return None
```

can be rewritten as:

```python3
def get_run_step(step: Step) -> str | None:
    if is_run_step(step):
        return step["run"]  # here, `step` is no longer just `Step` and has been type-narrowed to `RunStep`
    return None
```

One drawback of this is that it requires Python 3.10. I also do not know if there is any specific intention you have in mind to have this kind of API, so this is just a suggestion and up to you to decide.

Reply:
Python 3.13 introduces `TypeIs`, which appears to be recommended over `TypeGuard` in most cases. See https://peps.python.org/pep-0742/

Reply:
Nice. I did not know about `TypeIs`. Thanks, Nick.

Reply:
I didn't know about either of them. I have changed it to use `TypeGuard`, which does seem simpler. `TypeIs` does look better than `TypeGuard` (and in using `TypeGuard`, I did encounter the issue that `TypeIs` addresses where it doesn't narrow in the else block), but we currently have 3.11 as our min python version, so until we raise that we can't use `TypeIs`.